node/core/proposer/src/lib.rs

```diff
@@ -0,0 +1,187 @@
+use futures::select;
+use polkadot_node_subsystem::{messages::{AllMessages, ProvisionerInherentData, ProvisionerMessage}, SubsystemError};
+use polkadot_overseer::OverseerHandler;
+use polkadot_primitives::{
+	inclusion_inherent,
+	parachain::ParachainHost,
+	Block, Header,
+};
+use sc_block_builder::{BlockBuilderApi, BlockBuilderProvider};
+use sp_api::{ApiExt, ProvideRuntimeApi};
+use sp_blockchain::HeaderBackend;
+use sp_consensus::{Proposal, RecordProof};
+use sp_inherents::InherentData;
+use sp_runtime::traits::{DigestFor, HashFor};
+use sp_transaction_pool::TransactionPool;
+use std::{pin::Pin, sync::Arc, time};
+
+/// How long proposal can take before we give up and err out
+const PROPOSE_TIMEOUT: core::time::Duration = core::time::Duration::from_secs(2);
+
+/// Custom Proposer factory for Polkadot
+pub struct ProposerFactory<TxPool, Backend, Client> {
+	inner: sc_basic_authorship::ProposerFactory<TxPool, Backend, Client>,
+	overseer: OverseerHandler,
+}
+
+impl<TxPool, Backend, Client> ProposerFactory<TxPool, Backend, Client> {
+	pub fn new(
+		client: Arc<Client>,
+		transaction_pool: Arc<TxPool>,
+		overseer: OverseerHandler,
+	) -> Self {
+		ProposerFactory {
+			inner: sc_basic_authorship::ProposerFactory::new(
+				client,
+				transaction_pool,
+				None,
+			),
+			overseer,
+		}
+	}
+}
+
+impl<TxPool, Backend, Client> sp_consensus::Environment<Block>
+	for ProposerFactory<TxPool, Backend, Client>
+where
+	TxPool: 'static + TransactionPool<Block = Block>,
+	Client: 'static
+		+ BlockBuilderProvider<Backend, Block, Client>
+		+ ProvideRuntimeApi<Block>
+		+ HeaderBackend<Block>
+		+ Send
+		+ Sync,
+	Client::Api:
+		ParachainHost<Block> + BlockBuilderApi<Block> + ApiExt<Block, Error = sp_blockchain::Error>,
+	Backend:
+		'static + sc_client_api::Backend<Block, State = sp_api::StateBackendFor<Client, Block>>,
+	// Rust bug: https://github.com/rust-lang/rust/issues/24159
+	sp_api::StateBackendFor<Client, Block>: sp_api::StateBackend<HashFor<Block>> + Send,
+{
+	type CreateProposer = Pin<Box<
+		dyn Future<Output = Result<Self::Proposer, Self::Error>> + Send + 'static,
+	>>;
+	type Proposer = Proposer<TxPool, Backend, Client>;
+	type Error = Error;
+
+	fn init(&mut self, parent_header: &Header) -> Self::CreateProposer {
+		// create the inner proposer
+		let proposer = self.inner.init(parent_header).into_inner();
+
+		// clone the overseer handle (lightweight) so it can be moved into the future
+		let mut overseer = self.overseer.clone();
+
+		// we know that this function will be called at least once per proposed block,
+		// because this is where the parent header is supplied. Therefore, we can send
+		// the request for authorship data here.
+		let (sender, receiver) = futures::channel::oneshot::channel();
+		let parent_header_hash = parent_header.hash();
+
+		async move {
+			overseer.send_msg(AllMessages::Provisioner(
+				ProvisionerMessage::RequestInherentData(parent_header_hash, sender),
+			)).await?;
```

Comment:
Check my reasoning, please:

Awaiting here is safe because `overseer.send_msg` delegates to an internal `mpsc::Sender`, which allocates at least one buffer slot per sender. Cloning the overseer above cloned the sender, so this will never block, even if the overseer cannot dispatch the message for a long time.

Reply:
Yeah, that sounds right. However this property of the `OverseerHandler` isn't documented, so it doesn't seem stable to rely upon.

FWIW, we have a "fuck it" timeout somewhere in the proposer logic already. Could we bundle up all potentially-blocking (but realistically not blocking, or not long-blocking) work into a single future that we put under that timeout?

Reply:
d37e89e0